src/mbgl/map/tile_worker.cpp

```diff
@@ -1,73 +1,71 @@
-#include <mbgl/map/tile_parser.hpp>
-#include <mbgl/map/vector_tile_data.hpp>
-#include <mbgl/platform/log.hpp>
+#include <mbgl/text/collision_tile.hpp>
+#include <mbgl/map/tile_worker.hpp>
+#include <mbgl/style/style.hpp>
 #include <mbgl/style/style_layer.hpp>
-#include <mbgl/map/source.hpp>
-#include <mbgl/map/sprite.hpp>
+#include <mbgl/geometry/glyph_atlas.hpp>
 #include <mbgl/renderer/fill_bucket.hpp>
 #include <mbgl/renderer/line_bucket.hpp>
 #include <mbgl/renderer/symbol_bucket.hpp>
+#include <mbgl/platform/log.hpp>
 #include <mbgl/util/constants.hpp>
-#include <mbgl/style/style.hpp>
 
-#include <locale>
-
-namespace mbgl {
-
-// Note: This destructor is seemingly empty, but we need to declare it anyway
-// because this object has a std::unique_ptr<> of a forward-declare type in
-// its header file.
-TileParser::~TileParser() = default;
-
-TileParser::TileParser(const GeometryTile& geometryTile_,
-                       VectorTileData& tile_,
-                       const Style& style_,
-                       GlyphAtlas& glyphAtlas_,
-                       GlyphStore& glyphStore_,
-                       SpriteAtlas& spriteAtlas_,
-                       const util::ptr<Sprite>& sprite_)
-    : geometryTile(geometryTile_),
-      tile(tile_),
-      style(style_),
-      glyphAtlas(glyphAtlas_),
-      glyphStore(glyphStore_),
-      spriteAtlas(spriteAtlas_),
-      sprite(sprite_),
-      partialParse(false) {
-    assert(sprite);
+using namespace mbgl;
+
+TileWorker::TileWorker(const TileID& id_,
+                       Style& style_,
+                       const uint16_t maxZoom_,
+                       const std::atomic<TileData::State>& state_,
+                       std::unique_ptr<CollisionTile> collision_)
+    : style(style_),
+      id(id_),
+      maxZoom(maxZoom_),
+      state(state_),
+      collision(std::move(collision_)) {
+    assert(style.sprite);
 }
 
-bool TileParser::obsolete() const {
-    return tile.getState() == TileData::State::obsolete;
+TileWorker::~TileWorker() {
+    style.glyphAtlas->removeGlyphs(reinterpret_cast<uintptr_t>(this));
 }
 
-void TileParser::parse() {
-    for (const auto& layer_desc : style.layers) {
-        // Cancel early when parsing.
-        if (obsolete()) {
-            return;
-        }
+Bucket* TileWorker::getBucket(const StyleLayer& layer) const {
+    std::lock_guard<std::mutex> lock(bucketsMutex);
 
-        if (layer_desc->isBackground()) {
-            // background is a special, fake bucket
-            continue;
+    const auto it = buckets.find(layer.bucket->name);
+    if (it == buckets.end()) {
+        return nullptr;
+    }
+
+    assert(it->second);
+    return it->second.get();
+}
+
+size_t TileWorker::countBuckets() const {
```

Comment:
`countBuckets` is used to find out if, after reparsing a partial tile, anything has changed and we need to fire a "tile data changed" event. We can get rid of it after we map pending resources to tile data, because every reparse will always be caused by a resource that the tile data will be waiting for.